pkg/cluster/px/px.go

```diff
@@ -20,26 +39,199 @@ type Cluster interface {
 	return nil
 }
 
-func (p *pxCluster) Status(c *apiv1alpha1.Cluster) (*apiv1alpha1.ClusterStatus, error) {
+func (ops *pxClusterOps) Status(c *apiv1alpha1.Cluster) (*apiv1alpha1.ClusterStatus, error) {
+	logrus.Warnf("cluster status is not yet implemented")
 	return nil, nil
 }
 
-func (p *pxCluster) Upgrade(c *apiv1alpha1.Cluster) error {
-	logrus.Infof("upgrading px cluster: %s", c.Name)
+func (ops *pxClusterOps) Upgrade(new *apiv1alpha1.Cluster) error {
+	if len(new.Spec.PXVersion) == 0 {
+		return fmt.Errorf("new version of Portworx not given to upgrade API")
+	}
+
+	logrus.Infof("upgrading px cluster to %s", new.Spec.PXVersion)
+
+	// 1.  Start DaemonSet to download the new PX image and validate it completes
+	err := ops.runDockerPuller(new.Spec.PXVersion)
+	if err != nil {
+		return err
+	}
+
+	// 2. Find all pods that use PX shared volumes
+	// 3. Find controllers for these pods and drain them
+	//	For each controller in controllers
+	//		Add anti affinity so pods don't run
+	// 4. Wait till all px shared volumes are detached [ Implement in Status(...) ]
+	//		If timed out waiting, undo 3
+	// 5. Start rolling upgrade of PX DaemonSet
+	// 6. Wait till all PX pods are in ready state
+	// 7. [always] Undo 3
+
+	// DaemonSet
+	/*ds, err = ops.kubeClient.Extensions().DaemonSets(ds.Namespace).Update(ds)
+	if err != nil {
+		logrus.Errorf("failed updating daemonset: %s. Err: %s", ds.Name, err)
+		return err
+	}*/
 	return nil
 }
 
-func (p *pxCluster) Destroy(c *apiv1alpha1.Cluster) error {
-	logrus.Infof("destroying px cluster: %s", c.Name)
+func (ops *pxClusterOps) Destroy(c *apiv1alpha1.Cluster) error {
+	logrus.Warnf("destroying px cluster is implemented")
 	return nil
 }
 
-// NewPXClusterProvider creates a new PX cluster
-func NewPXClusterProvider(conf interface{}) (Cluster, error) {
-	return &pxCluster{}, nil
+// runDockerPuller runs the DaemonSet to start pulling the given image on all nodes
+func (ops *pxClusterOps) runDockerPuller(imageToPull string) error {
+	pullerName := "docker-puller"
+
+	labels := map[string]string{
+		"name": pullerName,
+	}
+	args := []string{"-i", imageToPull, "-w"}
+
+	var env []corev1.EnvVar
+	if len(ops.dockerRegistrySecret) != 0 {
+		logrus.Infof("using user-provided docker registry credentials from secret: %s", ops.dockerRegistrySecret)
+		env = append(env, corev1.EnvVar{
+			Name: "REGISTRY_USER",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: ops.dockerRegistrySecret,
+					},
+					Key: "REGISTRY_USER",
+				},
+			},
+		})
+
+		env = append(env, corev1.EnvVar{
+			Name: "REGISTRY_PASS",
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: ops.dockerRegistrySecret,
+					},
+					Key: "REGISTRY_PASS",
+				},
+			},
+		})
+	}
+
+	ds := &extensionsv1beta1.DaemonSet{
```

Comment:
Is this expecting to pull the image to the host? If so, I'm guessing that this should be run on privileged mode with access to docker socket, right? I do see `/var/run`, is that where the docker socket is?

Reply:
Yes, /var/run gives access to docker.sock. privileged mode is not required.